KAFKA-6975: Fix fetching from non-batch-aligned log start offset

apovzner opened the pull request:
It is possible that log start offset may fall in the middle of the batch after AdminClient#deleteRecords(). This will cause a follower starting from log start offset to fail fetching (all records). Use-cases when a follower will start fetching from log start offset includes: 1) new replica due to partition re-assignment; 2) new local replica created as a result of AdminClient#AlterReplicaLogDirs(); 3) broker that was down for some time while AdminClient#deleteRecords() move log start offset beyond its HW. 

Added two integration tests:
1) Produce and then AdminClient#deleteRecords() while one of the followers is down, and then restart of the follower requires fetching from log start offset;
2)  AdminClient#AlterReplicaLogDirs() after AdminClient#deleteRecords()

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

apovzner committed:
KAFKA-6975: Fix fetching from non-batch-aligned log start offset

apovzner committed:
Forgot to add new file

apovzner committed:
require message in UnexpectedAppendOffsetException

apovzner committed:
fixed abbreviations in test

apovzner committed:
Always throw UnexpectedAppendOffsetException if unexpected offset in append

ijuma commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -713,9 +713,19 @@ class Log(@volatile var dir: File,
+          if (appendInfo.firstOrLastOffset < nextOffsetMetadata.messageOffset) {
+            // we may still be able to recover in some situations
+            val firstOrLast = if (appendInfo.firstOffset.isDefined) "First" else "Last"
+            throw new UnexpectedAppendOffsetException(
```
Let's document this in the method scaladoc.

hachikuji replied:
I think the patch is primarily concerned with the case when the `firstOrLastOffset` is smaller than the current log start offset. I am wondering if we can handle this case here instead of in the caller. All of the logic in `LogManager.truncateFullyAndStartAt` is concerned with making the truncation safe for the cleaner, but if we are writing at the log start offset, then there won't be anything to clean anyway. Additionally, I don't think we actually want to change the log start offset in this case since we want it to match the leader.

I think we also need to be a little careful with the use of `firstOrLastOffset` here. For the new message format, this will be the first offset, so it is fine to use it for the base offset of the new segment that we create for this offset. But for the old message format, it will be the last offset of the batch (if compression is enabled). Maybe we have no choice but to decompress the batch in order to determine the first offset?

apovzner replied:
The logic in `LogManager.truncateFullyAndStartAt` also makes a checkpoint for all logs for the topic partition, which I think we should do, otherwise checkpoint will be inconsistent with actual data. 

We are setting log start offset to the leader's log start offset once we truncate fully and append. It is done in fetcher thread's processPartitionData().

We are not using `firstOrLastOffset` to find the first offset, we are using  `firstOffset` if it is defined. So, the fix is only for the new message format. Do we want to fix that for the old message format as well? The case of setting log start offset in the middle of the batch was introduced with deleteRecords in 1.0 (or 1.1), where we already had new message format -- meaning, that we don't need to backport this to much older versions. I guess it is possible that users could run new versions of kafka with older message format, but is that the case we want to support here?

apovzner replied:
@hachikuji Do you still think we should handle the truncation in append() vs. `LogManager.truncateFullyAndStartAt` given we also need to checkpoint (per my comment above)? Disregard my last paragraph about not addressing message format 1, I will add the code for finding first offset if it is not available.

hachikuji replied:
@apovzner Yeah, I was wondering whether the checkpoint was actually required. It could be, but my point was that we do not actually need to change the log start offset when this case is hit. We just need to be able to append a batch which starts a little earlier. In that case, I'm not sure checkpointing is actually needed (the checkpoints are just an optimization anyway).

Ah, I missed the use of `firstOffset`. I think we probably should try to handle the old message format as well since users have tended to upgrade the message format slowly. This should be a rare case, so I don't think decompressing to find the first offset is a big problem.

ijuma commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -713,9 +713,19 @@ class Log(@volatile var dir: File,
           // we are taking the offsets we are given
-          if (!appendInfo.offsetsMonotonic || appendInfo.firstOrLastOffset < nextOffsetMetadata.messageOffset)
+          if (!appendInfo.offsetsMonotonic)
             throw new IllegalArgumentException(s"Out of order offsets found in append to $topicPartition: " +
```
Should we be using a different exception here? For example, we should ideally have a test for this and it's easier to write if we can verify it by using the exception type instead of the exception message.

apovzner replied:
You mean we should add a new exception class for this? We have `InvalidOffsetException`, but this one could be thrown from same append() method if we couldn't append to the index, so I think we should not use that one. Maybe we should define `OffsetsOutOfOrderException`?

ijuma commented on core/src/main/scala/kafka/server/ReplicaAlterLogDirsThread.scala:
```diff
@@ -98,8 +99,23 @@ class ReplicaAlterLogDirsThread(name: String,
+          partition.appendRecordsToFutureReplica(records)
+        } else
+          throw e
+    }
```
Can we not share this code with the other class? Also, instead of `isDefined` followed by `get`, we should use pattern matching.

ijuma commented on core/src/test/scala/unit/kafka/log/LogTest.scala:
```diff
@@ -1894,6 +1894,24 @@ class LogTest {
+      log.appendAsFollower(records = batch)
+    } catch {
+      case e: UnexpectedAppendOffsetException =>
+        assertTrue(s"UnexpectedAppendOffsetException#firstOffset should be defined", e.firstOffset.isDefined)
```
Many lines above can be replaced by:

```scala
val actualFirstOffset = intercept[UnexpectedAppendOffsetException](log.appendAsFollower(records = batch).firstOffset
assertEquals(Some(firstOffset), actualFirstOffset)
```

ijuma commented on core/src/test/scala/unit/kafka/log/LogTest.scala:
```diff
@@ -1885,7 +1885,7 @@ class LogTest {
 
-  @Test(expected = classOf[IllegalArgumentException])
+  @Test(expected = classOf[UnexpectedAppendOffsetException])
   def testAppendWithOutOfOrderOffsetsThrowsException() {
```
Seems like we want to rename this method and add a test that checks for the true out of order case.

ijuma commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -23,6 +23,7 @@ import java.util.concurrent.{ExecutionException, TimeUnit}
 import java.io.File
 import java.util.concurrent.atomic.{AtomicBoolean, AtomicInteger}
 
+import kafka.api.Request
```
Do we really need this? The old Request classes are generally not used.

ijuma commented on core/src/main/scala/kafka/common/UnexpectedAppendOffsetException.scala:
```diff
@@ -0,0 +1,25 @@
+ * Indicates the follower or the future replica received an unexpected offset from the leader (or
+ * current replica)
+ */
+class UnexpectedAppendOffsetException(val message: String, val firstOffset: Option[Long]) extends RuntimeException(message) {
```
We should explain why `firstOffset` is optional.

ijuma reviewed (commented):
Thanks for the PR, this is a great catch. A few minor comments.

apovzner committed:
Using OffsetsOutOfOrderException for non-monotonic offsets in append, and added unit test

apovzner committed:
Also fixing for message format < 2

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -713,9 +713,26 @@ class Log(@volatile var dir: File,
+              s"Unexpected offset in append to $topicPartition. $firstOrLast " +
+              s"offset ${appendInfo.firstOrLastOffset} is less than the next offset ${nextOffsetMetadata.messageOffset}. " +
+              s"First 10 offsets in append: ${records.records.asScala.take(10).map(_.offset)}, last offset in" +
+              s" append: ${appendInfo.lastOffset}", firstOffset)
```
Perhaps it's useful to include logStartOffset in the message.

junrao commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -816,6 +816,87 @@ class AdminClientIntegrationTest extends IntegrationTestHarness with Logging {
+    client = AdminClient.create(createConfig)
+    sendRecords(producers.head, 100, topicPartition)
+
+    val result = client.deleteRecords(Map(topicPartition -> RecordsToDelete.beforeOffset(3L)).asJava)
```
It seems that there is no guarantee that the offset 3 is in the middle of a message set?

apovzner replied:
Yes, this is an integration test, but at least those tests are failing without my change. I just moved handling of the `UnexpectedAppendOffsetException` with truncation + append to Partition, so that there is no duplicate code in `ReplicaFetcherThread` and `ReplicaAlterLogDirsThread`. I can do a test that guarantees that log start offset is in the middle of the batch, but unfortunately there is no unit test for Partition class. Will try to create it, but not that simple (lots of mocking).

junrao commented on core/src/main/scala/kafka/common/UnexpectedAppendOffsetException.scala:
```diff
@@ -0,0 +1,26 @@
+package kafka.common
+
+/**
+ * Indicates the follower or the future replica received an unexpected offset from the leader (or
```
Perhaps we can be more specific on what unexpected offset means?

junrao reviewed (approved):
@apovzner : Thanks for the patch. LGTM. Just a few minor comments below.

apovzner committed:
Moved handling of UnexpectedAppendOffsetException to Partition#appendToFollower

apovzner committed:
Added Partition unit test

apovzner commented:
@junrao, @hachikuji and @ijuma Thanks for your reviews. I addressed all your comments. I moved handling of the UnexpectedAppendOffsetException to Partition, so that there is no duplicate code in ReplicaFetcherThread and ReplicaAlterLogDirsThread. We needed to handle it in the caller of Log#append, because we are moving log start and log end offset before the checkpointed recoveryPoint, so checkpoint is needed before the append.

junrao commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -545,15 +546,39 @@ class Partition(val topic: String,
 
-  def appendRecordsToFutureReplica(records: MemoryRecords) {
-    getReplica(Request.FutureLocalReplicaId).get.log.get.appendAsFollower(records)
+  private def doAppendRecordsToFollower(records: MemoryRecords, isFuture: Boolean): Unit = {
```
Perhaps this should be renamed to doAppendRecordsToFollowerOrFutureReplica? Ditto for appendRecordsToFollower() below.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -798,9 +800,26 @@ class Log(@volatile var dir: File,
+            throw new OffsetsOutOfOrderException(s"Out of order offsets found in append to $topicPartition: " +
+                                                 records.records.asScala.map(_.offset))
+
+          if (appendInfo.firstOrLastOffset < nextOffsetMetadata.messageOffset) {
```
This is an existing issue. I am wondering if we can tighten up the check here a bit. Currently, if records are of pre v2 format and have multiple batches, appendInfo.firstOrLastOffset returns the last offset of the last batch. In this case, it's better to use the last offset of the first batch for this check.

junrao reviewed (commented):
@apovzner : Thanks for the updated patch. Just a couple of more minor comments.

apovzner committed:
Using last offset of the first batch to check for unexpected offset

apovzner commented:
@junrao It makes sense to use last offset of the first batch vs. last offset for the check. I updated the PR accordingly. Thanks!

hachikuji commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -545,15 +546,39 @@ class Partition(val topic: String,
+    } catch {
+      case e: UnexpectedAppendOffsetException =>
+        val replica = if (isFuture) getReplica(Request.FutureLocalReplicaId).get else getReplica().get
+        if (replica.logEndOffset.messageOffset == replica.logStartOffset) {
```
Do you think we should validate that the unexpected records at least contain the expected log start offset?

apovzner replied:
Would that require uncompressing and looking at all the records? What do you think about checking if an expected offset is in between first and last offset of records?

hachikuji commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -545,15 +546,39 @@ class Partition(val topic: String,
-    getReplica(Request.FutureLocalReplicaId).get.log.get.appendAsFollower(records)
+  private def doAppendRecordsToFollower(records: MemoryRecords, isFuture: Boolean): Unit = {
+      if (isFuture)
+        getReplica(Request.FutureLocalReplicaId).get.log.get.appendAsFollower(records)
```
We don't necessarily need to fix it here since this was already a problem, but all these calls to `get` are discomforting. I'm wondering if we should have some helpers which raise (perhaps) `ReplicaNotAvailableException` if the option is empty.

apovzner replied:
That's true... I added Partition#getReplicaOrException and used it in `doAppendRecordsToFollower` + unit tests to verify this. I did not refactor any other code.

hachikuji commented on core/src/main/scala/kafka/common/OffsetsOutOfOrderException.scala:
```diff
@@ -0,0 +1,26 @@
+ * Indicates the follower received records with non-monotonically increasing offsets
+ */
+class OffsetsOutOfOrderException(message: String) extends RuntimeException(message) {
+  def this() = this(null)
```
Do we need this? Maybe we should always have a message?

hachikuji commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -811,6 +811,87 @@ class AdminClientIntegrationTest extends IntegrationTestHarness with Logging {
+    val leaders = createTopic(topic, numPartitions = 1, replicationFactor = serverCount)
+    val followerIndex = if (leaders(0) != servers(0).config.brokerId) 0 else 1
+
+    def waitForFollowerLSOandLEOAfterRestart(expectedLSO: Long, expectedLEO: Long): Unit = {
```
nit: maybe we can spell these out? Unfortunately "LSO" is also used for "last stable offset."

hachikuji commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -811,6 +811,87 @@ class AdminClientIntegrationTest extends IntegrationTestHarness with Logging {
+    restartDeadBrokers()
+    waitForFollowerLSOandLEOAfterRestart(expectedLSO=117L, expectedLEO=200L)
+
+    client.close()
```
nit: move to finally block so that a failure here won't cause other tests which are validating active threads to fail?

apovzner replied:
Just checked... I was wrong to call `client.close()` here, because client.close() is called on test's `tearDown` which is called after every test even if a test fails.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -84,13 +84,19 @@ case class LogAppendInfo(var firstOffset: Option[Long],
+  /**
+   * Get the first offset if it exists, else get the last offset of the first batch
+   */
+  def firstOrLastOfFirstBatchOffset: Long = firstOffset.getOrElse(lastOffsetOfFirstBatch)
```
The only other place that uses firstOrLastOffset is for maybeCacheTxnFirstOffsetMetadata(). I am wondering if that can/should also use firstOrLastOfFirstBatchOffset. Perhaps @hachikuji can comment on this.

hachikuji replied:
For transactions, we should be able to rely on having the first offset available, so using `firstOrLastOfFirstBatchOffset` should be fine. We could even use `firstOffset`, but that may require some reworking of the append logic.

By the way, I wonder if a friendlier name for this method might be `firstOffsetEstimate` or something like that?

hachikuji reviewed (commented):
Thanks for the patch. Left a few small comments.

apovzner committed:
addressed couple of comments

apovzner committed:
Added Partition#getReplicaOrException and tests

apovzner committed:
fixed typo

apovzner committed:
Append to log only if log start offset is within records' first and last offset

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -84,12 +84,15 @@ case class LogAppendInfo(var firstOffset: Option[Long],
                          validBytes: Int,
-                         offsetsMonotonic: Boolean) {
+                         offsetsMonotonic: Boolean,
+                         lastOffsetOfFirstBatch: Long) {
```
Need to update the parameter list above

hachikuji commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -811,6 +811,83 @@ class AdminClientIntegrationTest extends IntegrationTestHarness with Logging {
+    val leaders = createTopic(topic, numPartitions = 1, replicationFactor = serverCount)
+    val followerIndex = if (leaders(0) != servers(0).config.brokerId) 0 else 1
+
+    def waitForFollowerLogStartOffsetAndLEOAfterRestart(expectedLSO: Long, expectedLEO: Long): Unit = {
```
nit: This is still so verbose. How about this?
```scala
def waitForFollowerLog(expectedStartOffset: Long, expectedEndOffset: Long)
```

hachikuji commented on core/src/test/scala/unit/kafka/log/LogTest.scala:
```diff
@@ -1885,15 +1887,58 @@ class LogTest {
+      val actualFirstOffset = exception.firstOffset
+      val actualLastOffset = exception.lastOffset
+      assertEquals(s"Magic $magic. UnexpectedAppendOffsetException#firstOffset", firstOffset, actualFirstOffset)
+      assertEquals(s"Magic $magic. UnexpectedAppendOffsetException#lastOffset", firstOffset, actualLastOffset)
```
These assertions might be more useful if we had more than one record in the batch?

hachikuji commented on core/src/test/scala/unit/kafka/log/LogTest.scala:
```diff
@@ -1885,15 +1887,58 @@ class LogTest {
+    }
+  }
+
+  @Test(expected = classOf[UnexpectedAppendOffsetException])
```
nit: we're inconsistent on the user of `expected` and `assertThrows`.

hachikuji commented on core/src/test/scala/unit/kafka/log/LogTest.scala:
```diff
@@ -1885,15 +1887,58 @@ class LogTest {
+    val magicVals = Seq(RecordBatch.MAGIC_VALUE_V0, RecordBatch.MAGIC_VALUE_V1, RecordBatch.MAGIC_VALUE_V2)
+    for (magic <- magicVals) {
+      val batch = singletonRecordsWithLeaderEpoch(value = "random".getBytes, leaderEpoch = 1,
+                                                  offset = firstOffset, codec = CompressionType.LZ4, magicValue = magic)
```
Is it useful to verify behavior for non-compressed data (at least for the old message format)?

hachikuji commented on core/src/test/scala/unit/kafka/log/LogTest.scala:
```diff
@@ -1885,15 +1887,58 @@ class LogTest {
+  }
+
+  @Test(expected = classOf[UnexpectedAppendOffsetException])
+  def testAppendBelowExpectedOffsetThrowsException() {
```
Out of curiosity, why do we only verify this case with the new message format? The test case below seems more comprehensive.

apovzner replied:
This is an original test, I just changed its name. The more comprehensive test is a new test I added. But actually it's very easy to extend to multiple message formats. I will do.

hachikuji reviewed (commented):
Left a few additional comments. The fix looks good.

apovzner committed:
Better unit tests

hachikuji reviewed (approved):
Thanks, LGTM!
